tidy3d/plugins/autograd/functions.py

```diff
@@ -581,3 +582,57 @@ def morphological_gradient_external(
         The external morphological gradient of the input array.
     """
     return grey_dilation(array, size, structure, mode=mode, maxval=maxval) - array
+
+
+def rescale(
+    array: np.ndarray, out_min: float, out_max: float, in_min: float = 0.0, in_max: float = 1.0
+) -> np.ndarray:
+    """
+    Rescale an array from an arbitrary input range to an arbitrary output range.
+
+    Parameters
+    ----------
+    array : np.ndarray
+        The input array to be rescaled.
+    out_min : float
+        The minimum value of the output range.
+    out_max : float
+        The maximum value of the output range.
+    in_min : float, optional
+        The minimum value of the input range. Default is 0.0.
+    in_max : float, optional
+        The maximum value of the input range. Default is 1.0.
+
+    Returns
+    -------
+    np.ndarray
+        The rescaled array.
+    """
+    scaled = (array - in_min) / (in_max - in_min)
+    return scaled * (out_max - out_min) + out_min
+
+
+def threshold(
+    array: np.ndarray, vmin: float = 0.0, vmax: float = 1.0, level: Union[float, None] = None
```

Comment:
I think while this is more explicit, just `float = None` would have been ok too for us, just FYI

Reply:
Ideally we jump to Python >= 3.10 and can write `float | None` instead :grin:

Reply:
Maybe in late 2025 when python 3.9 is no longer supported? Unless there's a good argument for only supporting newer versions of python? I guess the `|` syntax would be quite nice 🤔 but probably will make some users unhappy to not support 3.8 or 3.9

Reply:
Yeah it's on the "at some point" list... I just really miss `|` for unions and `match`...